viewing_party/party.py

```diff
@@ -1,23 +1,156 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    if (not title) or (not genre) or (not rating):
+        return None
+    
+    movie_dict = {"title": title, "genre": genre, "rating": rating}
+    return movie_dict
+
+def add_to_watched(user_data, movie):
+    new_watched_list = user_data["watched"]
+    new_watched_list.append(movie)
+    user_data["watched"] = new_watched_list
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    new_watchlist_list = user_data["watchlist"]
+    new_watchlist_list.append(movie)
+    user_data["watchlist"] = new_watchlist_list
+    return user_data
+
+def watch_movie(user_data, title): 
+    for movie in list(user_data["watchlist"]):
+        if movie["title"] == title:
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
+            break
+
+    return user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    watched_movies_list = user_data["watched"]
+    if len(watched_movies_list) == 0:
+        return 0.0
+    sum_of_ratings = 0
+    for movie in watched_movies_list:
+        movie_rating = movie["rating"]
+        sum_of_ratings+= movie_rating
+    
+    return sum_of_ratings/len(watched_movies_list)
 
+def get_most_watched_genre(user_data):
+    watched_movies_list = user_data["watched"]
+    if len(watched_movies_list) == 0:
+        return None
+    #create a dictionary containing frequency of each genre
+    genre_freq = {}
+    for movie in watched_movies_list:
+        movie_genre = movie["genre"]
+        if movie_genre not in genre_freq:
+            genre_freq[movie_genre] = 1
+        else:
+            genre_freq[movie_genre] += 1
+        
+    #initially I used this code to find the most frequent genre using the values of the dictionary (which represents the freqnecy), and figuring out the key. But commented it out bc I found a one liner online to make it shorter.
+    #list_of_popular_genres = []
+    #max_frequency = 0
+    #for genre in genre_freq:
+    #    if genre_freq[genre] > max_frequency:
+    #        max_frequency = genre_freq[genre]
+    #        list_of_popular_genres.append(genre)
+    #return list_of_popular_genres[-1] 
+
+    return max(genre_freq, key=genre_freq.get)

```

Comment:
nicely done! If someone were to ask what this line is doing, would you be able to explain it?

Reply:
Explanation:
The python built-in max function accepts an optional parameter "key" which takes a function as its value, which is applied to each element before sorting, so that the elements are sorted based on the output of this function.

 Since the .get() method is used to return the value for a key, applying it to each key in the dictionary gives us the values of each key. Then the max function will be applied to the values of the dictionary genre_freq, instead of the keys. 

I believe that the equivalent lambda function would be:
max(genre_freq, key=lambda genre: genre_freq[genre])